app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/RedirectHelper.java

```diff
@@ -0,0 +1,43 @@
+package com.appsmith.server.helpers;
+
+import com.appsmith.server.constants.Security;
+import org.springframework.http.HttpHeaders;
+import org.springframework.util.StringUtils;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class RedirectHelper {
+
+    public static String DEFAULT_REDIRECT_URL = "/applications";
+    private static String REDIRECT_URL_HEADER = "X-Redirect-Url";
+
+    public static String getRedirectUrl(HttpHeaders httpHeaders) {
+        // First check if the custom redirect header is set
+        String redirectUrl = httpHeaders.getFirst(REDIRECT_URL_HEADER);
+
+        // If not, then try to get the header from Origin. We append
+        if (StringUtils.isEmpty(redirectUrl) && !StringUtils.isEmpty(httpHeaders.getOrigin())) {
+            redirectUrl = httpHeaders.getOrigin() + DEFAULT_REDIRECT_URL;
+        }
+
```

Comment:
There is an assumption here that the redirect URL coming from the header is always an absolute URL, and that the redirect URL in the `DEFAULT_REDIRECT_URL` constant is always a relative URL. I don't see much value in making those assumptions, so, I say why not support both?

```suggestion
        if (StringUtils.isEmpty(redirectUrl)) {
            redirectUrl = DEFAULT_REDIRECT_URL;
        }

        if (!(redirectUrl.startsWith("http://") || redirectUrl.startsWith("https://")) && !StringUtils.isEmpty(httpHeaders.getOrigin())) {
            redirectUrl = httpHeaders.getOrigin() + DEFAULT_REDIRECT_URL;
        }

```

I think this works better as now the redirect url, coming from any source, can be relative or absolute. When it's relative, and when an origin is available, we concatenate the two to get an absolute url. Now the client will have more flexibility in doing this.